engine/src/main/java/org/hibernate/validator/cfg/defs/SafeHtmlDef.java

```diff
@@ -6,16 +6,32 @@
  */
 package org.hibernate.validator.cfg.defs;
 
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
 import org.hibernate.validator.cfg.ConstraintDef;
 import org.hibernate.validator.constraints.SafeHtml;
+import org.hibernate.validator.internal.util.CollectionHelper;
+import org.hibernate.validator.internal.util.annotationfactory.AnnotationDescriptor;
+import org.hibernate.validator.internal.util.annotationfactory.AnnotationFactory;
 
 /**
  * @author Marko Bekta
  */
 public class SafeHtmlDef extends ConstraintDef<SafeHtmlDef, SafeHtml> {
 
+	private final Set<TagWithAttributes> tags;
```

Comment:
We discussed it with Gunnar. The idea would be to have a new ``AnnotationDef`` type (similar to the `ConstraintDef` approach with `annotationType` and a map of `parameters`).

This type would replace `TagWithAttributes` and `Attribute` and avoid the need to "refresh" the configuration after each operation.

The new ``AnnotationDef`` would be transformed to annotation proxies in `ConfiguredConstraint#createAnnotationProxy()` (e.g. you would need to special case if the value is an `AnnotationDef` and recurse into it).

The overall idea is:
 * to avoid unneeded instantiations
 * to keep the creation of the annotation proxies all in the same place

If it's not clear, feel free to ping me for additional discussions.